packages/std/src/query/mod.rs

```diff
@@ -62,6 +61,22 @@ pub enum QueryRequest<C> {
     #[cfg(feature = "stargate")]
     Ibc(IbcQuery),
     Wasm(WasmQuery),
+    Grpc(GrpcQuery),
+}
+
+/// Queries the chain using a grpc query.
+/// This allows to query information that is not exposed in our API.
+/// The chain needs to whitelist the supported queries.
+/// The drawback of this query is that you have to handle the protobuf encoding and decoding yourself.
+///
+/// The returned data is protobuf encoded. The protobuf type depends on the query.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct GrpcQuery {
+    /// The fully qualified service path used for routing,
+    /// eg. "custom/cosmos_sdk.x.bank.v1.Query/QueryBalance"
```

Comment:
Here we can provide a better example that makes it easier for users to find what they need. They look something like this:

```
$ grpcurl -plaintext -d '{ "interface_name": "cosmos.auth.v1beta1.AccountI" }' localhost:9090 cosmos.base.reflection.v1beta1.ReflectionService.ListImplementations
{
  "implementationMessageNames": [
    "/cosmos.vesting.v1beta1.PeriodicVestingAccount",
    "/cosmos.vesting.v1beta1.PermanentLockedAccount",
    "/cosmos.auth.v1beta1.BaseAccount",
    "/cosmos.auth.v1beta1.ModuleAccount",
    "/cosmos.vesting.v1beta1.BaseVestingAccount",
    "/cosmos.vesting.v1beta1.DelayedVestingAccount",
    "/cosmos.vesting.v1beta1.ContinuousVestingAccount"
  ]
}
$ !!
grpcurl -plaintext -d '{ "interface_name": "cosmos.auth.v1beta1.AccountI" }' localhost:9090 cosmos.base.reflection.v1beta1.ReflectionService.ListImplementations
{
  "implementationMessageNames": [
    "/cosmos.auth.v1beta1.BaseAccount",
    "/cosmos.auth.v1beta1.ModuleAccount",
    "/cosmos.vesting.v1beta1.BaseVestingAccount",
    "/cosmos.vesting.v1beta1.DelayedVestingAccount",
    "/cosmos.vesting.v1beta1.ContinuousVestingAccount",
    "/cosmos.vesting.v1beta1.PeriodicVestingAccount",
    "/cosmos.vesting.v1beta1.PermanentLockedAccount"
  ]
}
$  
```

Reply:
So, in this example "cosmos.base.reflection.v1beta1.ReflectionService.ListImplementations" would be the service path, right? I think I was just a bit confused by the translation explained in the ADR-21 doc linked in the stargate query docs.
Maybe a good example would be "cosmos.authz.v1beta1.Query.Grants". I think we could even add a reference to that cli tool to the docs.

Reply:
Here we have a real world example of a functional path:
https://github.com/jhernandezb/tokenfactory-example/blob/b091f5370de95a03c6be233b3e873a38d13c59ad/src/contract.rs#L179C18-L183

Reply:
In that case, there are some differences to the cli tool. It allows both `osmosis.tokenfactory.v1beta1.Query/DenomAuthorityMetadata` and `osmosis.tokenfactory.v1beta1.Query.DenomAuthorityMetadata`, but no leading `/`.

Reply:
Strange. Now I am officially stumped. But we should figure this out to ensure we don't create breaking changes that affect contracts which used one of the formats.

Reply:
Alright, I went through this with Pino. The stargate query we have right now calls into `GRPCQueryRouter::Route` to [get the handler for the message](https://github.com/CosmWasm/wasmd/blob/f29ae3543a0aaff5d4aa0ff2e72c8d7b779c143d/x/wasm/keeper/query_plugins.go#L347), which just looks it up in a map. Those routes are registered using [RegisterService](https://github.com/cosmos/cosmos-sdk/blob/v0.50.0-rc.1/baseapp/grpcrouter.go#L58) (this links to v0.50.0-rc.1), so that's where the `/osmosis.tokenfactory.v1beta1.Query/DenomAuthorityMetadata` format comes from.

Starting with the full sdk 0.50.0 release, [hybrid handlers were introduced](https://github.com/cosmos/cosmos-sdk/blob/v0.50.0/baseapp/grpcrouter.go#L77), which would look like `osmosis.tokenfactory.v1beta1.Query.DenomAuthorityMetadata`. These are separate from that `GRPCQueryRouter::Route` call though, so we would have to explicitly support them in wasmd.

Reply:
Here you see also an example format from when Terra disabled the query: https://github.com/terra-money/core/commit/2be170249147588f19b41c6daf4393b07ac5b009

`/cosmos.base.tendermint.v1beta1.Service/GetNodeInfo`

This query is obviously non-deterministic but shows the format that has been used ever since.